airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/stream.py

```diff
@@ -0,0 +1,467 @@
+                # this compares datatype of every column that the two schemas have in common
+                for col in column_superset:
+                    if (col in master_schema.keys()) and (col in this_schema.keys()) and (master_schema[col] != this_schema[col]):
+                        # if this column exists in a provided schema or schema state, we'll WARN here rather than throw an error
+                        # this is to allow more leniency as we may be able to coerce this datatype mismatch on read according to provided schema state
+                        # if not, then the read will error anyway
+                        if col in self._schema.keys():
+                            self.logger.warn(f"Detected mismatched datatype on column '{col}', in file '{fileclient._url}'. "
+                                             + f"Should be '{master_schema[col]}', but found '{this_schema[col]}'. "
+                                             + f"Airbyte will attempt to coerce this to {master_schema[col]} on read.")
+                        # else we're inferring the schema (or at least this column) from scratch and therefore throw an error on mismatching datatypes
+                        else:
+                            raise RuntimeError(
+                                f"Detected mismatched datatype on column '{col}', in file '{fileclient._url}'. "
+                                + f"Should be '{master_schema[col]}', but found '{this_schema[col]}'."
+                            )
+
+                # missing columns in this_schema doesn't affect our master_schema so we don't check for it here
+
+                # add to master_schema any columns from this_schema that aren't already present
+                for col, datatype in this_schema.items():
+                    if col not in master_schema.keys():
+                        master_schema[col] = datatype
+
+            self.logger.info(f"determined master schema: {master_schema}")
+            self.master_schema = master_schema
+
+        return self.master_schema
+
+    def stream_slices(
+        self, sync_mode: SyncMode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, Any]]]:
+        """
+        This builds full-refresh stream_slices regardless of sync_mode param.
+        1 file == 1 stream_slice.
+        Incremental stream_slices are implemented in the IncrementalFileStream child class.
+        """
+        
+        # TODO: this could be optimised via concurrent reads, however we'd lose chronology and need to deal with knock-ons of that
+        # we could do this concurrently both full and incremental by running batches in parallel
+        # and then incrementing the cursor per each complete batch
+        for last_mod, fileclient in self.time_ordered_fileclient_iterator():
+            yield [{
```

Comment:
Agree this seems odd but it's because when we're making incremental stream slices, they may contain more than one file (if multiple files have identical last_modified property). By yielding array here it means we only need to define read_records once and have it assume stream_slice is an array of object(s).